File: analyzers/structural/internal/metrics/type_metrics.go
package metrics

import (
	"sort"

	"slopslap.dev/structural/internal/facts"
)

func deepIf(statements []*facts.Statement, depth int, output *[]facts.Location) {
	for _, statement := range statements {
		if statement.Kind != facts.StmtIf {
			deepIf(statement.Body, depth, output)
			for _, branch := range statement.Cases {
				deepIf(branch.Body, depth, output)
			}
			continue
		}
		if depth+1 == 3 {
			*output = append(*output, statement.Location)
			continue
		}
		deepIf(statement.Body, depth+1, output)
		if len(statement.Else) == 1 && statement.Else[0].Kind == facts.StmtIf {
			deepIf(statement.Else, depth, output)
		} else {
			deepIf(statement.Else, depth+1, output)
		}
	}
}

func unique(values []string) []string {
	seen := make(map[string]struct{}, len(values))
	for _, value := range values {
		if value != "" {
			seen[value] = struct{}{}
		}
	}
	result := make([]string, 0, len(seen))
	for value := range seen {
		result = append(result, value)
	}
	sort.Strings(result)
	return result
}

func typeWMC(item *facts.Type) int {
	return typeWMCWith(item, Cyclomatic)
}

func typeWMCWith(item *facts.Type, cyclomatic func(*facts.Function) int) int {
	value := item.InterfaceMethodCount
	for _, method := range item.Methods {
		value += cyclomatic(method)
	}
	return value
}

func tcc(item *facts.Type) float64 {
	methods := make([]string, 0, len(item.MethodFields))
	for method := range item.MethodFields {
		methods = append(methods, method)
	}
	if len(methods) < 2 {
		return 0
	}
	methodsByField := make(map[string][]int)
	for index, method := range methods {
		for _, field := range item.MethodFields[method] {
			methodsByField[field] = append(methodsByField[field], index)
		}
	}
	connected := 0
	for left := 0; left < len(methods); left++ {
		connectedRights := make(map[int]struct{})
		for _, field := range item.MethodFields[methods[left]] {
			for _, right := range methodsByField[field] {
				if right > left {
					connectedRights[right] = struct{}{}
				}
			}
		}
		connected += len(connectedRights)
	}
	pairs := len(methods) * (len(methods) - 1) / 2
	return float64(connected) / float64(pairs)
}

File: analyzers/typescript/src/type-metrics.ts
import ts from "typescript";
import type { Measurement, SourceEntry, Subject } from "./model.js";
import {
  fieldUsageByMethod,
  isTypeDeclaration,
  methodLike,
  typeMemberCount,
  typeMembers,
  typeReferences,
  uniqueStrings,
  type FunctionFact,
} from "./surface.js";

function subject(source: ts.SourceFile, node: ts.Node, name: string): Subject {
  const start = source.getLineAndCharacterOfPosition(node.getStart(source));
  const end = source.getLineAndCharacterOfPosition(node.getEnd());
  return {
    name,
    symbol: name,
    start: {
      line: start.line + 1,
      column: start.character + 1,
      offset: node.getStart(source),
    },
    end: {
      line: end.line + 1,
      column: end.character + 1,
      offset: node.getEnd(),
    },
  };
}

function measurement(
  entry: SourceEntry,
  component: string,
  definition: string,
  scope: "function" | "file" | "type" | "expression",
  value: number,
  item: Subject,
  attributes: Record<string, unknown>,
): Measurement {
  return {
    unit_id: entry.unitId,
    component_id: component,
    definition_version: definition,
    path: entry.relativePath,
    scope,
    value,
    subject: item,
    attributes,
    provenance: {
      analyzer: "slopslap-typescript",
      analyzer_version: "0.1.0",
      rule: `${component}/${definition}`,
    },
  };
}

interface TypeMetricData {
  item: Subject;
  kind: "interface" | "class";
  wmc: number;
  tcc: number;
  foreignTypes: string[];
  foreignFields: string[];
}

function typeConnectivity(fields: Set<string>[]): number {
  const methodsByField = new Map<string, number[]>();
  for (let method = 0; method < fields.length; method++) {
    for (const field of fields[method] ?? []) {
      const owners = methodsByField.get(field);
      if (owners === undefined) methodsByField.set(field, [method]);
      else owners.push(method);
    }
  }
  let connected = 0;
  for (let left = 0; left < fields.length; left++) {
    const connectedRights = new Set<number>();
    for (const field of fields[left] ?? []) {
      for (const right of methodsByField.get(field) ?? []) {
        if (right > left) connectedRights.add(right);
      }
    }
    connected += connectedRights.size;
  }
  return connected;
}

function typeMetricData(
  entry: SourceEntry,
  statement: ts.ClassDeclaration | ts.InterfaceDeclaration,
  cyclomatic: (fact: FunctionFact) => number,
): TypeMetricData {
  const methods = typeMembers(statement, entry.sourceFile);
  const usage = methods.map((method) => fieldUsageByMethod(method));
  const fields = usage.map((item) => item.own);
  const pairs = (fields.length * (fields.length - 1)) / 2;
  const connected = typeConnectivity(fields);
  const tcc = pairs === 0 ? 0 : connected / pairs;
  const wmc =
    typeMemberCount(statement) +
    methods.reduce((sum, method) => sum + cyclomatic(method), 0);
  return {
    item: subject(
      entry.sourceFile,
      statement,
      statement.name?.text ?? "<anonymous-type>",
    ),
    kind: ts.isInterfaceDeclaration(statement) ? "interface" : "class",
    wmc,
    tcc,
    foreignTypes: uniqueStrings(typeReferences(statement)),
    foreignFields: uniqueStrings(usage.flatMap((item) => [...item.foreign])),
  };
}

function typeMetricMeasurementsForDeclaration(
  entry: SourceEntry,
  statement: ts.ClassDeclaration | ts.InterfaceDeclaration,
  requestedComponents: ReadonlySet<string>,
  cyclomatic: (fact: FunctionFact) => number,
): Measurement[] {
  const data = typeMetricData(entry, statement, cyclomatic);
  const output: Measurement[] = [];
  if (requestedComponents.has("cyclomatic_class_complexity"))
    output.push(
      measurement(
        entry,
        "cyclomatic_class_complexity",
        "pmd-v1",
        "type",
        data.wmc,
        data.item,
        { kind: data.kind },
      ),
    );
  if (requestedComponents.has("coupling_between_objects"))
    output.push(
      measurement(
        entry,
        "coupling_between_objects",
        "pmd-v1",
        "type",
        data.foreignTypes.length,
        data.item,
        { kind: data.kind, referenced_types: data.foreignTypes },
      ),
    );
  if (requestedComponents.has("god_class") && ts.isClassDeclaration(statement))
    output.push(
      measurement(
        entry,
        "god_class",
        "pmd-v1",
        "type",
        data.wmc,
        subject(
          entry.sourceFile,
          statement,
          statement.name?.text ?? "<anonymous-class>",
        ),
        {
          kind: "class",
          wmc: data.wmc,
          atfd: data.foreignFields.length,
          tcc: data.tcc,
        },
      ),
    );
  return output;
}

export function typeMetricMeasurements(
  entry: SourceEntry,
  requestedComponents: ReadonlySet<string>,
  cyclomatic: (fact: FunctionFact) => number,
): Measurement[] {
  const output: Measurement[] = [];
  for (const statement of entry.sourceFile.statements) {
    if (!isTypeDeclaration(statement)) continue;
    output.push(
      ...typeMetricMeasurementsForDeclaration(
        entry,
        statement,
        requestedComponents,
        cyclomatic,
      ),
    );
  }
  return output;
}
